<!--
University of Freiburg WS 2017/2018
Chair for Bioinformatics
Supervisor: Martin Raden
Author: Alexander Mattheis
-->

<div id="algorithm_description">
    <div class="description">
        The progressive alignment approach by
        <a href="https://doi.org/10.1007/BF02603120">Da-Fei Feng and Russell F. Doolittle (1987)</a>
        computes a multi-sequence-alignment (MSA) of a set of sequences based on pairwise alignments.
        This approximative approach identifies good MSA solutions in reasonable time.
        <br>
        <br>
	First, all pairwise alignments with maximal similarity of all input sequence pairs are computed.
        To this end, we use  <a href="https://doi.org/10.1016/0022-2836(82)90398-9">Osamu Gotoh's (1982)</a>
	approach available in 
	<a href="http://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Gotoh">our Gotoh implementation</a>.
        <br>
        In order to build a guide tree to generate the MSA, 
	observed similarity scores $S(a,b)$ of sequence pairs $a,b$ are converted into distances using
        $D(a,b) = -ln(S^{eff}_{a,b} \cdot f)$, where we set the scaling factor $f=1$ (originally $f = 100$). <br>
        The 'effective' similarity score $S^{eff}$ is computed by $S^{eff}_{a,b} = (S(a,b) - S^{rand}_{a,b})/(S^{max}_{a,b} - S^{rand}_{a,b})$
        where $S^{max}_{a,b} = (S(a,a)+S(b,b))/2$ provides the best alignment score for both sequences.
	$S^{rand}$ provides a kind of 'expected score' that can be obtained via sequence randomization. 
	Here, we use an alternative approach introduced in
	<a href="https://doi.org/10.1016/s0076-6879(96)66023-6">(Feng&amp;Doolittle, 1996)</a> 
	using the approximate formula for a given optimal alignment of $a$ and $b$:

        $$
        \quad
        S^{rand}_{a,b} =
        \frac{1}{L_{a,b}}
        \bigg(
        \sum_{x \in \bigcup_{a}} \sum_{y \in \bigcup_{b}}
        s(x,y) \cdot N_a(x) \cdot N_b(y)
        \bigg)
        + N_{a,b}^{e} \cdot \beta
        + N_{a,b}^{o} \cdot \alpha
        $$

        where <br>
        <div class="row">
            <div class="colW150">$L_{a,b}$</div>
            <div class="colW400">number of columns in the alignment of $a,b$</div>
	    </div>

        <div class="row">
            <div class="colW150">$\bigcup_{a}, \bigcup_{b}$</div>
            <div class="colW400">characters present in $a$ and $b$</div>
        </div>

        <div class="row">
            <div class="colW150">$s(x,y)$</div>
            <div class="colW400">similarity-score for aligning the characters $x$ and $y$</div>
        </div>

        <div class="row">
            <div class="colW150">$N_a(x), N_b(y)$</div>
            <div class="colW400">number of characters $x$ and $y$ in sequences $a$ and $b$, resp.</div>
        </div>

        <div class="row">
            <div class="colW150">$N_{a,b}^{e}$</div>
            <div class="colW400">number of gaps (enlargements)</div>
        </div>

        <div class="row">
            <div class="colW150">$N_{a,b}^{o}$</div>
            <div class="colW400">number of gap openings (adaptation for Gotoh)</div>
        </div>

        <div class="row">
            <div class="colW150">$\alpha, \beta$</div>
            <div class="colW400">
                parameters from
                <a href="http://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Gotoh">Gotoh</a> algorithm
            </div>
        </div>

        The resulting distances are then used
        to create a phylogenetic tree that guides the generation of the MSA, i.e. more related (similar) sequences are fused first.
        Here, the UPGMA algorithm envisioned
        by <a href="https://archive.org/details/cbarchive_33927_astatisticalmethodforevaluatin1902">Robert R. Sokal and Charles D. Michener (1958)</a> is applied.
        <br>
	    The tree defines the order in which sequences or groups of sequences
        have to be fused to create a multi-sequence-alignment (MSA). 
	    The fusion is done based on the best pairwise alignment between members of the groups to be aligned.
        To simplify the fusion, the 'once a gap, always a gap'-strategy is applied, 
	    i.e. once a gap is introduced in a subalignment it is preserved within all subsequent fusions.
        To this end, gaps in subalignments are replaced with 
        a neutral symbol $\#$ before aligning them (i.e. $s(\cdot,\#) = s(\#,\cdot) = 0$).
        <br>
        <br>
        For a given input, the stated computation steps generate
        <ul>
            <li>pairwise alignments</li>
            <li>distance matrix</li>
            <li>phylogenetic tree</li>
            <li>MSA via fusions</li>
        </ul>
        and are represented in reverse order below. <br>
    </div>

    <div class="picture">
        <img src="Feng-Doolittle-120x90.png">
    </div>
</div>

<h2 class="header">Input:</h2>
<div id="algorithm_input">
    <!-- ko foreach: $root.input.sequences() -->
        <div class="row">
            <div class="colW100">
                <label>
                    Sequence <span data-bind="text: $root.input.clusterNames()[$index()]"></span>:
                </label>
            </div>

            <div class="colW175">
                <input class="sequence" data-bind="value: $data" placeholder="EXAMPLE 'ATC'" type="text">

                <!-- ko if: $index() == $root.input.clusterNames().length-1 -->
                    <div class="group_hint">
                        <b>Hint:</b> <br>
                        The "+" and "-" buttons can be used to add or remove sequences.
                        <br>
                        <br>
                        <b>Hint 2:</b> <br>
                        Duplicates of sequences <br>
                        are excluded from computations.
                        <!-- This is a bigger code readability and runtime enhancement:
                        sequences are stored as keys in hash-tables and
                        two times the same key is not possible in a hash-table! -->
                    </div>
                <!-- /ko -->
            </div>

            <!-- ko if: $index() == $root.input.clusterNames().length-1 -->
                <button class="add_remove" data-bind="click: $root.input.addRow" title="Add Sequence" type="button">+</button>
            <!-- /ko -->

            <!-- ko if: $index() == $root.input.clusterNames().length-1 && $root.input.clusterNames().length > 2 -->
                <button class="add_remove" data-bind="click: $root.input.removeRow" title="Remove Sequence" type="button">-</button>
            <!-- /ko -->
        </div>
    <!-- /ko -->

    <div class="row">
		<div class="colW100"><label>Scoring in $s$:</label></div>

		<div class="colW400">
            <span class="group"> <!-- Microsoft Browsers will fallback on text-fields using following input type -->
                Match <input class="fx_parameter" data-bind="value: input.match" id="match" type="number">
                Mismatch <input class="fx_parameter" data-bind="value: input.mismatch" id="mismatch" type="number">
            </span>
		</div>
    </div>

    <div class="row">
		<div class="colW100"><label>Scoring in $g$:</label></div>

		<div class="colW400">
            <span class="group"> <!-- Microsoft Browsers will fallback on text-fields using following input type -->
                Gap opening $\alpha$ <input class="fx_parameter" data-bind="value: input.baseCosts" id="base_costs" type="number">
                Enlargement $\beta$ <input class="fx_parameter" data-bind="value: input.enlargement" id="enlargement" type="number">
            </span>
            <br>
            <br>
            <span data-bind="text: $root.input.gapFunction"></span>

            <div class="group_hint">
				<b>Hint:</b> <br>
				For similarity maximization, <br> match scores should be positive and all other scores lower. <br>
			</div>
		</div>
    </div>
</div>

<h2 class="header">Output:</h2>

<div id="algorithm_output">
    <div class="output">
        <div class="final">
            <table class="final_result_header">
                <thead>
					<tr>
						<th>
                            Final result
						</th>
					</tr>
                </thead>
            </table>

            <div class="final_result_with_scrollbar">
                <table class="final_result">
                    <tbody>
                        <tr>
                            <td class="entry entry_start">
                                <!-- ko foreach: $root.output.progressiveAlignment() -->
                                    <code data-bind="text: $root.output.progressiveAlignment()[$index()]"></code> <br>
                                <!-- /ko -->
                            </td>
                        </tr>
                    </tbody>
                </table>
            </div>
        </div>

        <div class="intermediate_result_header">
            <h4>Intermediate Results</h4>
        </div>

        <div class="merge_steps">
            <table class="result_header">
                <thead>
					<tr>
						<th>
                            Merge steps
						</th>
					</tr>
                </thead>
            </table>

            <div class="result_with_scrollbar">
                <table class="result_categories">
                    <thead>
                        <tr>
                            <th>
                                <small><b>Group 1</b></small>
                            </th>

                            <th>
                                <small><b>Group 2</b></small>
                            </th>

                            <th>
                                <small><b>Guide Alignment</b></small>
                            </th>

                            <th>
                                 <small><b>Result</b></small>
                            </th>
                        </tr>
                    </thead>

                    <tbody>
                        <!-- ko foreach: $root.output.guideAlignments() -->
                            <tr>
                                <td class="entry entry_start">
                                    <!-- ko foreach: $root.output.firstGroups()[$index()] -->
                                        <code data-bind="text: $data"></code> <br>
                                    <!-- /ko -->
                                </td>

                                <td class="entry entry_start">
                                    <!-- ko foreach: $root.output.secondGroups()[$index()] -->
                                        <code data-bind="text: $data"></code> <br>
                                    <!-- /ko -->
                                </td>

                                <td class="entry entry_start">
                                    <code data-bind="text: $root.output.guideAlignments()[$index()][0]"></code> <br>
                                    <code data-bind="text: $root.output.guideAlignments()[$index()][1]"></code> <br>
                                    <code data-bind="text: $root.output.guideAlignments()[$index()][2]"></code>
                                </td>

                                 <td class="entry entry_start">
                                     <code data-bind="text: $root.output.joinedGroupNames()[$index()]"></code>:
                                     <!-- ko foreach: $root.output.joinedGroups()[$index()] -->
                                        <!-- ko if: $index() == 0 -->
                                            <code data-bind="text: $data"></code> <br>
                                        <!-- /ko -->

                                        <!-- ko if: $index() != 0 -->
                                            <code>  </code><code data-bind="text: $data"></code> <br>
                                        <!-- /ko -->
                                     <!-- /ko -->
                                </td>
                            </tr>
                         <!-- /ko -->
                    </tbody>
                </table>
            </div>
        </div>

        <div class="intermediate_result_header">
            <h4 class="subresult">Phylogenetic Tree</h4>
        </div>

        <div class="newick_tree">
            <table class="result_header">
                <thead>
					<tr>
						<th>
                            Newick Tree
						</th>
					</tr>
                </thead>
            </table>

            <div class="result_with_scrollbar">
                <table class="result">
                    <tbody>
                        <tr>
                            <td class="entry entry_start">
                                <code data-bind="text: $root.output.newickString()"></code>
                            </td>
                        </tr>
                    </tbody>
                </table>
            </div>
        </div>

        <div class="tree_container"> <!-- allows to delete and reinsert the div -->
            <div id="phylogenetic_tree"></div>
        </div>

        <div class="group_hint">
            <b>Visualization done with</b> <br>
            Smits SA, Ouverney CC, 2010. jsPhyloSVG: <br>
            A Javascript Library for Visualizing Interactive and Vector-Based Phylogenetic Trees on the Web. <br>
            <a href="http://www.plosone.org/article/info:doi/10.1371/journal.pone.0012267">
                PLoS ONE 5(8): e12267. doi:10.1371/journal.pone.0012267
            </a>
        </div>

        <div class="intermediate_result_header">
            <h4 class="subresult">Distance Table</h4>
        </div>

        <div class="distance_table">
            <table class="calculation">
                <thead>
					<tr>
						<th></th>
                        <!-- ko foreach: $root.output.remainingClusters()[0] -->
                            <!-- ko if: $index() < $root.output.distanceMatrix()[0].length -->
                                <th data-bind="drawChar: [$data, undefined]"></th>
                            <!-- /ko -->
                        <!-- /ko -->
					</tr>
                </thead>

                <tbody>
					<!-- ko foreach: $root.output.distanceMatrix()[0]--> <!-- to get i-indexes = $parentContext.$index() -->
						<tr>
                            <th data-bind="drawChar: [$root.output.remainingClusters()[0][$index()], undefined]"></th>

                            <!-- ko foreach: $root.output.distanceMatrix()[0] --> <!-- to get j-indexes = $index() -->
                                    <td class="non_selectable_entry"
                                        data-bind="text: $root.output.distanceMatrix()[$parentContext.$index()][$index()]"></td>
                            <!-- /ko -->
						</tr>
                    <!-- /ko -->
                </tbody>
            </table>
        </div>

        <button data-bind="click: $root.output.toggleVisibility">UPGMA Iterations</button>

        <div class="matrices" data-bind="visible: $root.output.showMatrices">
            <!-- ko foreach: $root.output.distanceMatrices() -->
                <!-- ko if: $index() > 0 -->
                    <div class="intermediate_result_header">
                        <h4 class="header subresult">Iteration <span data-bind="text: $index()"></span>:</h4>
                    </div>

                    <div class="distance_table">
                        <table class="calculation">
                            <thead>
                                <tr>
                                    <th></th>
                                    <!-- ko foreach: $root.output.remainingClusters()[$index()] -->
                                        <th data-bind="drawChar: [$data, undefined]"></th>
                                    <!-- /ko -->
                                </tr>
                            </thead>

                            <tbody>
                                <!-- ko foreach: $root.output.distanceMatrices()[$index()] --> <!-- to get i-indexes = $parentContext.$index() -->
                                    <tr>
                                        <th data-bind="drawChar: [$root.output.remainingClusters()[$parentContext.$index()][$index()], undefined]"></th>
                                        <!-- ko foreach: $root.output.distanceMatrices()[$parentContext.$index()][$index()] --> <!-- to get j-indexes = $index() -->
                                            <td class="non_selectable_entry"
                                                data-bind="text: $root.output.distanceMatrices()[$parentContext.$parentContext.$index()][$parentContext.$index()][$index()]">
                                            </td>
                                        <!-- /ko -->
                                    </tr>
                                <!-- /ko -->
                            </tbody>
                        </table>
                    </div>
                <!-- /ko -->
            <!-- /ko -->
        </div>

        <div class="intermediate_result_header">
            <h4 class="subresult">Pairwise Data</h4>
        </div>

        <div class="pairwise_data">
            <div class="result_with_scrollbar">
                <table class="result_categories top_border">
                    <thead>
                        <tr>
                            <th>
                                <small><b>Sequence Pair</b></small>
                            </th>

                            <th>
                                <small><b>Length $L$</b></small>
                            </th>

                            <th>
                                <small><b>Score $S$</b></small>
                            </th>

                            <th>
                                <small><b>Gaps $N^e$</b></small>
                            </th>

                            <th>
                                <small><b>Gap openings $N^o$</b></small>
                            </th>
                        </tr>
                    </thead>

                    <tbody>
                        <!-- ko foreach: $root.output.sequencePairNames() -->
                            <tr>
                                <td class="entry entry_start">
                                    <code>(</code><code data-bind="text: $data[0]"></code><code>,</code>
                                    <code data-bind="text: $data[1]"></code><code>)</code>
                                    <br>
                                </td>

                                <td class="entry entry_start">
                                    <code data-bind="text: $root.output.alignmentLengths()[$index()]"></code> <br>
                                </td>

                                <td class="entry entry_start">
                                    <code data-bind="text: $root.output.similarities()[$index()]"></code> <br>
                                </td>

                                <td class="entry entry_start">
                                    <code data-bind="text: $root.output.gapNumbers()[$index()]"></code> <br>
                                </td>

                                <td class="entry entry_start">
                                    <code data-bind="text: $root.output.gapStarts()[$index()]"></code> <br>
                                </td>
                            </tr>
                        <!-- /ko -->
                    </tbody>
                </table>
            </div>
        </div>
    </div>
</div>
