<!--
University of Freiburg WS 2017/2018
Chair for Bioinformatics
Supervisor: Martin Raden
Authors: Alexander Mattheis, Martin Raden
-->

<div id="algorithm_description">
    <div class="description"> <!-- what-why-how description -->
        <a href="http://dx.doi.org/10.1006/jmbi.2000.4042">C&eacute;dric Notredame, Desmond G. Higgins and Jaap Heringa</a>
        introduced 2000 an approach that
        allows a very accurate computation of multiple sequence alignments (MSA),
        the Tree-based Consistency Objective Function for AlignmEnt Evaluation (short: <em>T-coffee</em>).
        It's a tree-based approach,
        because it uses a progressive alignment strategy
        under a consistency-based function
        (tree can be computed with <a href="http://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Agglomerative-Clustering">UPGMA</a>).
        The advantage to traditional approaches like
        <a href="https://doi.org/10.1007/BF02603120">Feng-Doolittle (1987)</a>
        is that the results are not based on single pairwise alignments for a 
	    set of input sequences $\mathbb{S}$,
        but rather on an arbitrary large pool of pairwise alignments on $\mathbb{S}$ from
	    which a position specific scoring function is derived.
        This so called extended library $EL^{a,b}_{i,j}$ provides for each
        pair of sequences $a,b \in \mathbb{S}$ and respective positions $i,j$ a score.
        <br />
        <br />
	    To this end, first a primary library $L^{a,b}_{i,j}$ is computed.
        It covers the information how often a certain pair of sequence positions
	    $a_i$ and $b_j$ were aligned within the set of input alignments. This is
	    weighted by the 'accuracy' of the respective alignments $\mathcal{A}(a,b)$
        based on sequence identity:

          $$
        \quad
        L^{a,b}_{i,j} = \sum_{\mathcal{A}(a,b)}\begin{cases}
        weight \big(\mathcal{A}(a,b) \big)
        & \text{if } \binom{a_i}{b_j} \in \mathcal{A}(a,b)\\
        0 & \text{else}
        \end{cases}
        $$

        where $L^{a,b}_{i,j} = L^{b,a}_{j,i}$ and

	    $$
        \quad
	    weight \big(\mathcal{A}(a,b) \big)
        = \frac {matches \big(\mathcal{A}(a,b) \big) \cdot 100}
                {matches \big(\mathcal{A}(a,b) \big) + mismatches \big(\mathcal{A}(a,b) \big)} \text{.}
        $$

        Subsequently, the extended library $EL^{a,b}_{i,j}$ is calculated to gather consistency information, 
	    i.e. 'indirect alignments' of $a_i$ with $b_j$ when both are aligned to some $c_k$
	    (while taking all possible sequences $c$ into account).
        Thus, all sequence triples from $\mathbb{S}$ are considered, which leads to the following formula:

        $$
        \quad
        EL^{a,b}_{i,j}
        = L^{a,b}_{i,j} + \sum_{c \in \mathbb{S} \setminus \{a,b\}} \sum_{1 \leq k \leq|c|} min \big(L^{a,c}_{i,k}, L^{c,b}_{k,j} \big).
        $$

        Finally, this scoring function $EL$ is used in a progressive alignment scheme (see
        <a href="http://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Feng-Doolittle">Feng-Doolittle</a>)
        to compute a multiple sequence alignment.
        Unlike in the <a href="https://doi.org/10.1007/BF02603120">Feng-Doolittle (1987)</a> approach,
        the whole columns (aligned positions of all sequences) of two groups are used to 
	    score/compute their joint alignment.
        To this end, the average alignment score of two columns is used
        (all chars of first column aligned with all chars of second column) 
	    while gaps are ignored by setting $\alpha = \beta = 0$.
	    The latter is based on the idea that the library already indirectly
	    reflects the gap scoring that was present in the underlying input alignments.
        <br />
        <br />
        In the algorithm, it is possible to add information
        from arbitrary alignments to emphasize particularly important
        positions of aligned sequences.
        Here, you can use (multiple optimal) global and local alignments of the input sequences
	    using the <a href="http://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Gotoh">Gotoh</a> algorithm.
        Each alignment is used in the formula from above to compute $L$.
        <br />
        <br />
        For a given input, the computation steps are
        <ul>
            <li>pairwise alignments</li>
            <li>primary and extended library</li>
            <li>phylogenetic tree</li>
            <li>MSA via progressive alignment</li>
            <li>multiple sequence alignment (MSA) with SoP score</li>
        </ul>
        which are represented in reverse order below.
        The sum-of-pairs (SoP) score of the final MSA is simply the score sum of all pairwise
        alignments within the MSA, i.e. the global alignment scores of pairwise row 
	combinations where gap-only columns are removed.
    </div>

    <div class="picture">
        <img src="Notredame-Higgins-Heringa-120x90.png" />
    </div>
</div>

<h1>Input:</h1>
<div id="algorithm_input">
    <!-- ko foreach: $root.input.sequences() -->
        <div class="row">
            <div class="colW100">
                <label>
                    Sequence <span data-bind="text: $root.input.clusterNames()[$index()]"></span>:
                </label>
            </div>

            <div class="colW225">
                <input class="sequence_multi" data-bind="value: $data" placeholder="EXAMPLE 'ATC'" title="Allowed are A-Z and '-'." type="text">

                <!-- ko if: $index() == $root.input.clusterNames().length-1 -->
                    <div class="group_hint">
                        <b>Hint:</b> <br />
                        The "+" and "-" buttons can be used to add or remove sequences.
                        Duplicates of sequences <br />
                        are excluded from computations.
                        <!-- The distance formula
                        is not defined for two sequences of same type,
                        because the denominator gets zero.
                        Also, this allows a bigger code readability and runtime enhancement:
                        Sequences are stored as keys in hash-tables and
                        two times the same key is not possible in a hash-table!
                        -->
                    </div>
                <!-- /ko -->
            </div>

            <!-- ko if: $index() == $root.input.clusterNames().length-1 -->
                <button class="add_remove" data-bind="click: $root.input.addRow" title="Add Sequence" type="button">+</button>
            <!-- /ko -->

            <!-- ko if: $index() == $root.input.clusterNames().length-1 && $root.input.clusterNames().length > 2 -->
                <button class="add_remove" data-bind="click: $root.input.removeRow" title="Remove Sequence" type="button">-</button>
            <!-- /ko -->
        </div>
    <!-- /ko -->

    <div class="row">
        <div class="colW100"><label>Alignments:</label></div>

        <div class="colW400">
            max. # 
            <input class="fx_parameter" data-bind="value: input.globalAlignmentsPerSequencePair"
                            id="global_alignments_per_sequence_pair" type="number">
            optimal alignment(s) per sequence pair
        </div>
    </div>

    <div class="row">
		<div class="colW100"><label>Scoring in $s$:</label></div>

		<div class="colW400">
            <span class="group"> <!-- Microsoft Browsers will fallback on text-fields using following input type -->
                Match <input class="fx_parameter" data-bind="value: input.match" id="match" type="number">
                Mismatch <input class="fx_parameter" data-bind="value: input.mismatch" id="mismatch" type="number">
            </span>
		in similarity scoring
		</div>
    </div>

    <div class="row">
		<div class="colW100"><label>Scoring in $g$:</label></div>

		<div class="colW400">
            <span class="group"> <!-- Microsoft Browsers will fallback on text-fields using following input type -->
                Gap opening $\alpha$ <input class="fx_parameter" data-bind="value: input.baseCosts" id="base_costs" type="number">
                Enlargement $\beta$ <input class="fx_parameter" data-bind="value: input.enlargement" id="enlargement" type="number">
            </span>
            <br />
            <br />
            <span data-bind="text: $root.input.gapFunction"></span>

            <div class="group_hint">
				<b>Hint:</b> <br />
				For similarity maximization, <br /> match scores should be positive and all other scores lower. <br />
			</div>
		</div>
    </div>

    <div class="row">
        <div class="colW100"><label>Add local library</label></div>
        <div class="colW400"><input data-bind="checked: $root.input.useLocalLibrary" type="checkbox"></div>
    </div>

    <div class="localLibraryParameters" data-bind="visible: $root.input.useLocalLibrary">
        <h2 class="input_heading">Local Library Parameters</h2>

        <div class="row">
            <div class="colW100"><label>Alignments:</label></div>

            <div class="colW400">
                <div class="inner_row">
                    <div class="column">
                        total number alignments
                    </div>

                    <div class="colW50">
                        <input data-bind="enable: input.localAlignmentsPerSequencePair() == 1, value: input.totalNumberAlignments"
                               id="total_number_alignments" type="number"> <br />
                    </div>
                </div>

                 <div class="inner_row">
                     <div class="column">
                         max. # <input class="fx_parameter" data-bind="value: input.localAlignmentsPerSequencePair"
                                       id="local_alignments_per_sequence_pair" type="number">
                         optimal alignment(s) per sequence pair
                     </div>
                </div>
            </div>
        </div>

        <div class="row">
            <div class="colW100"><label>Scoring in $s$:</label></div>

            <div class="colW400">
                <span class="group"> <!-- Microsoft Browsers will fallback on text-fields using following input type -->
                    Match <input class="fx_parameter" data-bind="value: input.matchLocal" id="match_local" type="number">
                    Mismatch <input class="fx_parameter" data-bind="value: input.mismatchLocal" id="mismatch_local" type="number">
                </span>
            </div>
        </div>

        <div class="row">
            <div class="colW100"><label>Scoring in $g$:</label></div>

            <div class="colW400">
                <span class="group"> <!-- Microsoft Browsers will fallback on text-fields using following input type -->
                    Gap opening $\alpha$ <input class="fx_parameter" data-bind="value: input.baseCostsLocal" id="base_costs_local" type="number">
                    Enlargement $\beta$ <input class="fx_parameter" data-bind="value: input.enlargementLocal" id="enlargement_local" type="number">
                </span>
                <br />
                <br />
                <span data-bind="text: $root.input.gapFunctionLocal"></span>

                <div class="group_hint">
                    <b>Hint:</b> <br />
                    For similarity maximization, <br /> match scores should be positive and all other scores lower. <br />
                </div>
            </div>
        </div>
    </div>
</div>

<h1>Output:</h1>
<div id="algorithm_output">
    <div class="final">
        <table class="final_result_header">
            <thead>
                <tr>
                    <th>
                        Final result
                    </th>
                </tr>
            </thead>
        </table>

        <div class="final_result_with_scrollbar">
            <table class="final_result">
                <tbody>
                    <tr>
                        <td class="entry entry_start">
                            <!-- ko foreach: $root.output.progressiveAlignment() -->
                                <code data-bind="text: $root.output.progressiveAlignment()[$index()]"></code> <br />
                            <!-- /ko -->
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>

        <table class="final_result_footer">
            <tr>
                <th>
                    <small>
                        Affine SoP score: <span data-bind="text: $root.output.score()"></span>
                    </small>
                </th>
            </tr>
        </table>
    </div>

    <h2>Intermediate Results</h2>

    <div class="merge_steps">
        <table class="result_header">
            <thead>
                <tr>
                    <th>
                        Fusion steps
                    </th>
                </tr>
            </thead>
        </table>

        <div class="result_with_scrollbar">
            <table class="result_categories">
                <thead>
                    <tr>
                        <th>
                            <small><b>Group 1</b></small>
                        </th>

                        <th>
                            <small><b>Group 2</b></small>
                        </th>

                        <th>
                             <small><b>Result</b></small>
                        </th>
                    </tr>
                </thead>

                <tbody>
                    <!-- ko foreach: $root.output.firstGroups() -->
                        <tr>
                            <td class="entry entry_start">
                                <code data-bind="text: $root.output.firstGroupsNames()[$index()]"></code>: <br />
                                <!-- ko foreach: $root.output.firstGroups()[$index()] -->
                                    <code data-bind="text: $data"></code> <br />
                                <!-- /ko -->
                            </td>

                            <td class="entry entry_start">
                                <code data-bind="text: $root.output.secondGroupsNames()[$index()]"></code>: <br />
                                <!-- ko foreach: $root.output.secondGroups()[$index()] -->
                                    <code data-bind="text: $data"></code> <br />
                                <!-- /ko -->
                            </td>

                            <td class="entry entry_start">
                                <code data-bind="text: $root.output.joinedGroupNames()[$index()]"></code>: <br />
                                <!-- ko foreach: $root.output.joinedGroups()[$index()] -->
                                    <code data-bind="text: $data"></code> <br />
                                <!-- /ko -->
                            </td>
                        </tr>
                    <!-- /ko -->
                </tbody>
            </table>
        </div>
    </div>

    <h2>Phylogenetic Tree</h2>
    <div class="newick_tree">
        <table class="result_header">
            <thead>
                <tr>
                    <th>
                        Newick Tree
                    </th>
                </tr>
            </thead>
        </table>

        <div class="result_with_scrollbar">
            <table class="result">
                <tbody>
                    <tr>
                        <td class="entry entry_start">
                            <code data-bind="text: $root.output.newickString()"></code>
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>

    <div class="tree_container"> <!-- allows to delete and reinsert the div -->
        <div id="phylogenetic_tree"></div>
    </div>

    <!-- ko if: $root.output.newickString().length !== 1 && $root.output.newickString().indexOf(SYMBOLS.MINUS) === -1 -->
        <div class="group_hint">
            <b>Visualization done with</b> <br />
            Smits SA, Ouverney CC, 2010. jsPhyloSVG: <br />
            A Javascript Library for Visualizing Interactive and Vector-Based Phylogenetic Trees on the Web. <br />
            <a href="http://www.plosone.org/article/info:doi/10.1371/journal.pone.0012267">
                PLoS ONE 5(8): e12267. doi:10.1371/journal.pone.0012267
            </a>
        </div>
    <!-- /ko -->

    <div class="group_hint">
        <b>Hint: </b> Visualization is only shown, if there are no negative distances.
    </div>

    <h2>Libraries</h2>
    <div class="pairwise_data">
        <div class="result_with_scrollbar">
            <table class="result_categories top_border">
                <thead>
                    <tr>
                        <th>
                            <small><b>Sequence Pair</b></small>
                        </th>

                        <th>
                            <small><b>Position Pair</b></small>
                        </th>

                        <th>
                            <small><b>Primary Library $L$</b></small>
                        </th>

                        <th>
                            <small><b>Extended Library $EL$</b></small>
                        </th>
                    </tr>
                </thead>

                <tbody>
                    <!-- ko foreach: $root.output.sequencePairsNames() -->
                        <tr>
                            <td class="entry entry_start">
                                <code>(</code><code data-bind="text: $data[0]"></code><code>,</code>
                                <code data-bind="text: $data[1]"></code><code>)</code>
                                <br />
                            </td>

                            <td class="entry entry_start">
                                <!-- ko foreach: $root.output.libPositionPairs()[$index()] -->
                                    <code>(</code><code data-bind="text: $data"></code><code>)</code> <br />
                                <!-- /ko -->
                            </td>

                            <td class="entry entry_start">
                                <!-- ko foreach: $root.output.primLibValues()[$index()] -->
                                    <code data-bind="text: $data"></code> <br />
                                <!-- /ko -->
                            </td>

                            <td class="entry entry_start">
                                <!-- ko foreach: $root.output.extendedLibValues()[$index()] -->
                                    <code data-bind="text: $data"></code> <br />
                                <!-- /ko -->
                            </td>
                        </tr>
                    <!-- /ko -->
                </tbody>
            </table>
        </div>

        <table class="pairwise_data_footer">
            <tr>
                <th>
                    <small>
                        <b>Hint:</b> Only sequence pairs with alignment length bigger zero displayed!
                    </small>
                </th>
            </tr>
        </table>
    </div>

    <h2>Alignments</h2>
     <div class="pairwise_data">
        <div class="result_with_scrollbar">
            <table class="result_categories top_border">
                <thead>
                    <tr>
                        <th>
                            <small><b>Sequence Pair</b></small>
                        </th>

                        <th>
                            <small><b>Global Alignments</b></small>
                        </th>

                        <th>
                            <small><b>Local Alignments</b></small>
                        </th>
                    </tr>
                </thead>

                <tbody>
                    <!-- ko foreach: $root.output.sequencePairsNames() -->
                        <tr>
                            <td class="entry entry_start">
                                <code>(</code><code data-bind="text: $data[0]"></code><code>,</code>
                                <code data-bind="text: $data[1]"></code><code>)</code>
                                <br />
                            </td>

                            <td class="entry entry_start">
                                <!-- ko foreach: $root.output.alignmentsGlobal()[$index()] -->
                                    <code data-bind="text: $data[0]"></code> <br />
                                    <code data-bind="text: $data[1]"></code> <br />
                                    <code data-bind="text: $data[2]"></code> <br />
                                    <br />
                                <!-- /ko -->
                            </td>

                            <td class="entry entry_start">
                                <!-- ko foreach: $root.output.alignmentsLocal()[$index()] -->
                                    <code data-bind="text: $data[0]"></code> <br />
                                    <code data-bind="text: $data[1]"></code> <br />
                                    <code data-bind="text: $data[2]"></code> <br />
                                    <br />
                                <!-- /ko -->
                            </td>
                        </tr>
                    <!-- /ko -->
                </tbody>
            </table>
        </div>
    </div>
</div>
